### ISSUE TITLE
pls gab code

### ISSUE THREAD

**Issue by wassname** [rarefin/SEQ_VCR#1]
please, I'd like to see a reference example of this loss

**Comment by wassname** [comment]
pls

**Comment by wassname** [comment]
```py
from einops import rearrange, reduce, repeat
from jaxtyping import Float, Int
from typing import Tuple, List, Union, Optional, Dict
from torch import Tensor
import torch
from collections import defaultdict
import torch.nn as nn


def calc_distribution_loss(
    hs: Float[Tensor, "b t h"], chunk_size: int = 32
) -> Float[Tensor, ""]:
    B, T, P = hs.shape
    total_loss = 0.0
    eps = 1e-6

    # hs = F.layer_norm(hs)

    x = rearrange(hs, "b t h -> t b h")

    # Get distribution per timestep
    x_dist = F.softmax(x, dim=-1)

    # Compute uniform target
    uniform = torch.ones_like(x_dist) / P

    x_dist_log = torch.log(x_dist + eps)

    # KL divergence from uniform (prevent collapse)
    kl_loss = F.kl_div(x_dist_log, uniform, reduction="batchmean", log_target=False)

    # Entropy term (maximize feature usage)
    entropy = -(x_dist * x_dist_log).sum(-1).mean()

    total_loss = kl_loss - 0.1 * entropy

    return total_loss / T

def calc_seq_vcr_loss(hs: Float[Tensor, "b t h"], η = 1e-4) -> Float[Tensor, ""]:
    B, T, P = hs.shape

    # Compute covariance per timestep across batch
    x = rearrange(hs, "b t h -> t b h")  # Shape: (T, B, P)

    # Calculate mean and center per timestepx.dtype
    x_mean = x.mean(dim=1, keepdim=True).detach()  # Shape: (T, 1, P)
    x_centered = x - x_mean

    # Compute covariance matrices for each timestep
    C = torch.bmm(x_centered.transpose(1, 2), x_centered) / (B - 1)  # Shape: (T, P, P)

    # Setup mask for diagonal elements
    diag = torch.eye(P, dtype=torch.bool, device=x.device).detach()

    # Calculate losses
    # TODO move these to function arguments. And try to make the loss automatically balanced if I can
    λ1, λ2 = 1. / 5000, 1. / 50000
    
    # The Variance Term encourages unit variance in each dimension (most important)
    var_loss = torch.relu(1 - torch.sqrt(C * diag + η))
    var_loss = reduce(var_loss, "t h1 h2 -> ", "sum") / (T * diag.sum())

    # the Covariance Term penalizes covariance between different dimensions, promoting decorrelation and diversity in representations
    non_diag = (~diag).detach()
    cov_loss = (C * non_diag).pow(2)
    cov_loss = reduce(cov_loss, "t h1 h2 -> ", "sum") / (T * non_diag.sum())

    # Combine and reduce
    loss = λ1 * var_loss + λ2 * cov_loss
    return loss, {"loss_vcr_var": var_loss.item(), "loss_vcr_cov": cov_loss.item()}

class VCRLoss(nn.Module):
    def __init__(self, H=1536, D=256):
        # TODO set these param properly
        super().__init__()
        self.down_proj = nn.Linear(
            H, D, bias=False
        )  # FIXME is this meant to be learnable?
        nn.init.orthogonal_(self.down_proj.weight)
        self.down_proj.weight.requires_grad = False
        # FIXME we should be applying this to the final high dim output too, but wait for reference implementation https://github.com/rarefin/SEQ_VCR

    def forward(self, hs_l: Float[Tensor, "l b t h"]) -> Float[Tensor, "b"]:
        # for each layer
        loss = 0
        logs = defaultdict(int)
        for hs in hs_l:
            hs = self.down_proj(hs)
            # loss += calc_distribution_loss(hs)
            loss_i, extra = calc_seq_vcr_loss(hs)
            for k, v in extra.items():
                logs[k] += v
            loss += loss_i
            torch.cuda.empty_cache()

        return loss, dict(logs)

```

I had a go, but there are questions that I could not answer from reading the paper

**Comment by wassname** [comment]
pls

**Comment by wassname** [comment]
but seriously, why not release the code, even messy but working code will reveal implementation details that the paper does not, and allow people to replicate your work

**Comment by wassname** [comment]
this is my attempt at the code, but it's not yet tested for performance
```py
from einops import rearrange, reduce, repeat
from jaxtyping import Float, Int
from typing import Tuple, List, Union, Optional, Dict
from torch import Tensor
import torch
from collections import defaultdict
import torch.nn as nn
import torch.nn.functional as F


def calc_seq_vcr_loss(
    hs: Float[Tensor, "b t h"],
    η: float = 1e-3,
    λ1: float = 0.001,
    λ2: float = 0.0001,
) -> Tuple[Float[Tensor, ""], Dict[str, float]]:
    """
    Fully-vectorized Seq-VCR loss: no Python for‐loop.


    Usage:
    class Model(nn.Module):
        def __init__(self, ...):
            ...
            self.vcr_loss = VCRLoss(H=hidden_size, D=proj_dim)
            ...

        def forward(self, x):
            loss = super().forward(x)
            ...
            if self.training and self.config.loss_seq_vcr:
                loss_vcr, logs_vcr = self.vcr_loss(hs_l)
                loss += loss_vcr

            ...
            return loss, logits, logs

    """
    B, T, P = hs.shape
    # (T, B, P)
    x = rearrange(hs, "b t h -> t b h")
    x_center = x - x.mean(dim=1, keepdim=True)  # center per timestep

    # full covariance tensor: (T, P, P)
    C = torch.einsum("t b p, t b q -> t p q", x_center, x_center) / (B - 1)

    # variance term: sqrt(diag + η)
    diag = torch.diagonal(C, dim1=1, dim2=2)             # (T, P)
    var_loss = F.relu(1 - (diag + η).sqrt()).sum() / (T * P)

    # covariance term: mask out diag, square‐&‐sum
    mask = ~torch.eye(P, device=hs.device, dtype=torch.bool)  # (P, P)
    cov_loss = (C[..., mask].pow(2)).sum() / (T * P * (P - 1)) # (P, P)
    # Note this is different from the original code, which normalizes by T * P. But this means we don't have to keep changing λ2 like in the paper, so it seems better to me. The downside is we don't regularise large hidden spaces more, but changing λ2 shows that didn't work anyway.

    # normalize
    loss = (λ1 * var_loss + λ2 * cov_loss)
    return loss, {
        "loss_vcr_var": λ1 * var_loss.item(),
        "loss_vcr_cov": λ2 * cov_loss.item(),
    }

class VCRLoss(nn.Module):
    def __init__(self, H=1536, D=32):
        super().__init__()

        # Fixed random projection
        self.down_proj = nn.Linear(
            H, D, bias=False
        )
        nn.init.orthogonal_(self.down_proj.weight)
        self.down_proj.weight.requires_grad = False
        # FIXME we should be applying this to the final high dim output too, but wait for reference implementation https://github.com/rarefin/SEQ_VCR

    def forward(self, hs_l: Float[Tensor, "l b t h"]) -> Float[Tensor, "b"]:
        # for each layer
        loss = 0
        logs = defaultdict(float)
        for hs in hs_l:
            hs2 = self.down_proj(hs)
            loss_i, extra = calc_seq_vcr_loss(hs2)
            for k, v in extra.items():
                logs[k] += v
            loss += loss_i
            torch.cuda.empty_cache()

        return loss, dict(logs)

```

**Comment by yongding-tao** [comment]
It's interesting to see while Yann LeCun is one of the authors, the repo is still in a 'code coming soon' state after eight months.
